src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "CommandPalette.h"
+
+#include "CommandPalette.g.cpp"
+
+using namespace winrt;
+using namespace winrt::TerminalApp;
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::System;
+using namespace winrt::Windows::Foundation;
+
+namespace winrt::TerminalApp::implementation
+{
+    CommandPalette::CommandPalette()
+    {
+        InitializeComponent();
+
+        _filteredActions = winrt::single_threaded_observable_vector<winrt::TerminalApp::Command>();
+        _allActions = winrt::single_threaded_vector<winrt::TerminalApp::Command>();
+
+        if (CommandPaletteShadow())
+        {
+            // Hook up the shadow on the command palette to the backdrop that
+            // will actually show it. This needs to be done at runtime, and only
+            // if the shadow actually exists. ThemeShadow isn't supported below
+            // version 18362.
+            CommandPaletteShadow().Receivers().Append(ShadowBackdrop());
+            // "raise" the command palette up by 16 units, so it will cast a shadow.
+            Backdrop().Translation({ 0, 0, 16 });
+        }
+    }
+
+    // Method Description:
+    // - Toggles the visibility of the command palette. This will auto-focus the
+    //   input box within the palette.
+    // - TODO GH#TODO: When we add support for commandline mode, accept a parameter here
+    //   for which mode we should enter in.
+    void CommandPalette::ToggleVisibility()
+    {
+        const bool isVisible = Visibility() == Visibility::Visible;
+        if (!isVisible)
+        {
+            // Become visible
+            Visibility(Visibility::Visible);
+            _SearchBox().Focus(FocusState::Programmatic);
```

Comment:
So, this one is curious. If they entered through the keyboard, it might make sense to set this to ::Keyboard. If we switch to using the focus model like searchbox, we might be able to automatically funnel focus in in the right mode.

Reply:
("switch to using the focus model ..." was the subject of my comment in the header, "i wish we didn't need to put CommandPalette in charge of its own focusing")